roachtest: hotspotsplits/nodes=4 failed [inconsistency]

[(roachtest).hotspotsplits/nodes=4 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2776621&tab=buildLog) on [master@e09b93fe62541c3a94f32a723778660b528a0792](https://github.com/cockroachdb/cockroach/commits/e09b93fe62541c3a94f32a723778660b528a0792):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/hotspotsplits/nodes=4/run_1
	cluster.go:1754,test_runner.go:862: r38 (/Table/53/1/-182447479829764077) is inconsistent: RANGE_INCONSISTENT 
		(1) attached stack trace
		  -- stack trace:
		  | main.(*cluster).CheckReplicaDivergenceOnDB
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:1705
		  | main.(*cluster).FailOnReplicaDivergence.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:1751
		  | github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:140
		  | main.(*cluster).FailOnReplicaDivergence
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:1748
		  | main.(*testRunner).runTest
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:862
		  | main.(*testRunner).runWorker
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:472
		  | main.(*testRunner).Run.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:275
		  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:351
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		Wraps: (2) r38 (/Table/53/1/-182447479829764077) is inconsistent: RANGE_INCONSISTENT 
		  | (n3,s3):3: checksum 099280635e0ad488ea26cb5f253acebde95ceaa12eb3a8088e41ce92866507e5a664046b208ad370517d6af5c839c135dd1c27e612bbb0d59f75357a06084b61 [minority]
		  | - stats: contains_estimates:0 last_update_nanos:1615788916721166939 intent_age:0 gc_bytes_age:0 live_bytes:301241874 live_count:1149 key_bytes:28077 key_count:1149 val_bytes:301213797 val_count:1149 intent_bytes:0 intent_count:0 separated_intent_count:0 sys_bytes:702 sys_count:6 abort_span_bytes:0 
		  | - stats.Sub(recomputation): last_update_nanos:1615788916721166939 live_bytes:301241874 live_count:1149 key_bytes:28077 key_count:1149 val_bytes:301213797 val_count:1149 sys_bytes:702 sys_count:6 
		  | (n2,s2):2: checksum ff3cbc427622cc2ebf8ae40b42f37870f5662226a33bd636662a3423b268d19c15ae26c4f10181e9a80e531d00bb5127cbbcd44c21088fe8e5745d10653687fa
		  | - stats: contains_estimates:0 last_update_nanos:1615788916721166939 intent_age:0 gc_bytes_age:0 live_bytes:300979696 live_count:1148 key_bytes:28052 key_count:1148 val_bytes:300951644 val_count:1148 intent_bytes:0 intent_count:0 separated_intent_count:0 sys_bytes:702 sys_count:6 abort_span_bytes:0 
		  | - stats.Sub(recomputation): last_update_nanos:1615788916721166939 live_bytes:300979696 live_count:1148 key_bytes:28052 key_count:1148 val_bytes:300951644 val_count:1148 sys_bytes:702 sys_count:6 
		  | (n1,s1):1: checksum ff3cbc427622cc2ebf8ae40b42f37870f5662226a33bd636662a3423b268d19c15ae26c4f10181e9a80e531d00bb5127cbbcd44c21088fe8e5745d10653687fa
		  | - stats: contains_estimates:0 last_update_nanos:1615788916721166939 intent_age:0 gc_bytes_age:0 live_bytes:300979696 live_count:1148 key_bytes:28052 key_count:1148 val_bytes:300951644 val_count:1148 intent_bytes:0 intent_count:0 separated_intent_count:0 sys_bytes:702 sys_count:6 abort_span_bytes:0 
		  | - stats.Sub(recomputation): last_update_nanos:1615788916721166939 live_bytes:300979696 live_count:1148 key_bytes:28052 key_count:1148 val_bytes:300951644 val_count:1148 sys_bytes:702 sys_count:6
		Error types: (1) *withstack.withStack (2) *errutil.leafError

```

<details><summary>More</summary><p>


Artifacts: [/hotspotsplits/nodes=4](https://teamcity.cockroachdb.com/viewLog.html?buildId=2776621&tab=artifacts#/hotspotsplits/nodes=4)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Ahotspotsplits%2Fnodes%3D4.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

@sumeerbhola, did we break `cockroach debug keys <store>`?

```
$ roachprod ssh tobias-1617031837-01-n4cpu4:2
$ ~/cockroach debug keys /mnt/data1/cockroach/auxiliary/checkpoints/r59_at_2866/
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1  a panic has occurred!
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +panic: iterator with constraint=2 is being used with key /Min that has constraint=1
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +(1) attached stack trace
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  -- stack trace:
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | runtime.gopanic
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /usr/local/go/src/runtime/panic.go:969
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).checkConstraint
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:309
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).SeekGE
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:241
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/storage.iterateOnReader
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /go/src/github.com/cockroachdb/cockroach/pkg/storage/engine.go:1059
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/storage.(*Pebble).MVCCIterate
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble.go:711
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.runDebugKeys
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /go/src/github.com/cockroachdb/cockroach/pkg/cli/debug.go:223
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.MaybeDecorateGRPCError.func1
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /go/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:194
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/spf13/cobra.(*Command).execute
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:826
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/spf13/cobra.(*Command).ExecuteC
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:914
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/spf13/cobra.(*Command).Execute
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:864
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.Run
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:311
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.doMain
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:112
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.Main
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:60
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | main.main
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /go/src/github.com/cockroachdb/cockroach/pkg/cmd/cockroach/main.go:26
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | runtime.main
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /usr/local/go/src/runtime/proc.go:204
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | runtime.goexit
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  |        /usr/local/go/src/runtime/asm_amd64.s:1374
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +Wraps: (2) panic: iterator with constraint=2 is being used with key /Min that has constraint=1
E210329 21:28:35.364276 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +Error types: (1) *withstack.withStack (2) *errutil.leafError
panic: iterator with constraint=2 is being used with key /Min that has constraint=1 [recovered]
        panic: iterator with constraint=2 is being used with key /Min that has constraint=1 [recovered]
        panic: iterator with constraint=2 is being used with key /Min that has constraint=1

goroutine 1 [running]:
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.RecoverAndReportPanic(0x5a01120, 0xc0000760a8, 0xc000cc0000)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:90 +0x9f
panic(0x43989c0, 0xc002ab38f0)
        /usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc0003d6080, 0x5a01120, 0xc0000760a8)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x31e
panic(0x43989c0, 0xc002ab38f0)
        /usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).checkConstraint(0xc002a50620, 0x0, 0x0, 0x0, 0x1bf8600)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:309 +0x2c5
github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).SeekGE(0xc002a50620, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:241 +0x491
github.com/cockroachdb/cockroach/pkg/storage.iterateOnReader(0x5a6c820, 0xc002ab38d0, 0x0, 0x0, 0x0, 0x7c99bec, 0x2, 0x2, 0x0, 0xc000d64e60, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/engine.go:1059 +0x249
github.com/cockroachdb/cockroach/pkg/storage.(*Pebble).MVCCIterate(0xc001783450, 0x0, 0x0, 0x0, 0x7c99bec, 0x2, 0x2, 0x0, 0xc000d64e60, 0x203000, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble.go:711 +0x125
github.com/cockroachdb/cockroach/pkg/cli.runDebugKeys(0x7c71c80, 0xc000164760, 0x1, 0x1, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/debug.go:223 +0x21f
github.com/cockroachdb/cockroach/pkg/cli.MaybeDecorateGRPCError.func1(0x7c71c80, 0xc000164760, 0x1, 0x1, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:194 +0x79
github.com/spf13/cobra.(*Command).execute(0x7c71c80, 0xc000164740, 0x1, 0x1, 0x7c71c80, 0xc000164740)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:826 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0x7c6ed00, 0xc0000760a8, 0xc00169402a, 0xa)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:914 +0x30b
github.com/spf13/cobra.(*Command).Execute(...)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:864
github.com/cockroachdb/cockroach/pkg/cli.Run(...)
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:311
github.com/cockroachdb/cockroach/pkg/cli.doMain(0x7c71c80, 0xc00169402a, 0xa, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:112 +0x11a
github.com/cockroachdb/cockroach/pkg/cli.Main()
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:60 +0x13a
main.main()
        /go/src/github.com/cockroachdb/cockroach/pkg/cmd/cockroach/main.go:26 +0x25
```

Ditto for range-keys, but that doesn't seem pebble related.
```
$ ~/cockroach debug range-data /mnt/data1/cockroach/auxiliary/checkpoints/r59_at_2866 59
panic: field raftpb.HardState.term has invalid type: got uint64, want pointer [recovered]
        panic: field raftpb.HardState.term has invalid type: got uint64, want pointer [recovered]
        panic: field raftpb.HardState.term has invalid type: got uint64, want pointer
```

Only one failure so far, but I think it's due to random I/O problems, but maybe still worth a look for storage

```
panic: runtime error: index out of range [8677] with length 8617

goroutine 127024 [running]:
panic(0x491f960, 0xc00bc40e80)
        /usr/local/go/src/runtime/panic.go:1064 +0x545 fp=0xc0081f5508 sp=0xc0081f5440 pc=0x48b285
runtime.goPanicIndex(0x21e5, 0x21a9)
        /usr/local/go/src/runtime/panic.go:88 +0xa5 fp=0xc0081f5550 sp=0xc0081f5508 pc=0x488d85
github.com/cockroachdb/pebble/internal/manifest.(*L0Sublevels).InitCompactingFileInfo(0xc006b5fa00, 0xc003840730, 0x1, 0x1)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/internal/manifest/l0_sublevels.go:372 +0x505 fp=0xc0081f58
10 sp=0xc0081f5550 pc=0x16366c5
github.com/cockroachdb/pebble.(*DB).removeInProgressCompaction(0xc000b07400, 0xc003ae9b80)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/compaction.go:1217 +0x4ca fp=0xc0081f5c20 sp=0xc0081f5810 
pc=0x1694f2a
github.com/cockroachdb/pebble.(*DB).compact1(0xc000b07400, 0xc003ae9b80, 0x0, 0x5965d00, 0xc01751c600)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/compaction.go:1878 +0x514 fp=0xc0081f5e90 sp=0xc0081f5c20 
pc=0x1698fb4
github.com/cockroachdb/pebble.(*DB).compact.func1(0x5a00f00, 0xc000654240)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/compaction.go:1814 +0x92 fp=0xc0081f5f00 sp=0xc0081f5e90 p
c=0x16eaff2
runtime/pprof.Do(0x5a00f00, 0xc000654240, 0xc000483340, 0x1, 0x1, 0xc001599f98)
        /usr/local/go/src/runtime/pprof/runtime.go:40 +0xcc fp=0xc0081f5f68 sp=0xc0081f5f00 pc=0xd32e2c
github.com/cockroachdb/pebble.(*DB).compact(0xc000b07400, 0xc003ae9b80, 0x0)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/compaction.go:1811 +0xa5 fp=0xc0081f5fc8 sp=0xc0081f5f68 p
c=0x1698a85
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc0081f5fd0 sp=0xc0081f5fc8 pc=0x4c46c1
created by github.com/cockroachdb/pebble.(*DB).maybeScheduleCompactionPicker
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/compaction.go:1607 +0x411
```

```
[25783.274975] print_req_error: critical medium error, dev nvme0n1, sector 571228680
[25783.343170] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25783.360659] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25785.153934] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25785.166182] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25785.311798] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25785.338059] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25785.437281] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25785.446931] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25785.581243] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25788.485226] print_req_error: 31 callbacks suppressed
[25788.485228] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25788.494437] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25788.632245] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25788.641283] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25788.770603] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25788.809604] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25789.022107] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25789.031900] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25792.939138] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25792.950418] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25793.510220] print_req_error: 6 callbacks suppressed
[25793.510221] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25793.521429] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25793.657254] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25793.666236] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25793.828371] print_req_error: critical medium error, dev nvme0n1, sector 571229128
[25793.866891] print_req_error: critical medium error, dev nvme0n1, sector 571229128
```

Repros are coming in fast. I hope I didn't break the assertion while "improving" it.

```
TBG CRASH DETECTED GOID=191: r41: truncated cache to [1927, 2026] but got an entry for 2027..2046
add=false bytesGuessed=0bytesAdded=0 entriesAdded=0 bytesRemoved=0 entriesRemoved=0
panic: TBG CRASH DETECTED GOID=191: r41: truncated cache to [1927, 2026] but got an entry for 2027..2046
add=false bytesGuessed=0bytesAdded=0 entriesAdded=0 bytesRemoved=0 entriesRemoved=0
```

Time to poke at the core dump.

Run 27 on n3 (tobias-1617787468-37-n4cpu4:3)

```
panic: TBG CRASH DETECTED GOID=136: TBG term regression: idx 598 at term=7 appended with logterm=9
```

run 71, n3 `tobias-1617787468-13-n4cpu4:3`
```
panic: TBG CRASH DETECTED GOID=152: r39: truncated cache to [1154, 1253] but got an entry for 1254..1268
add=false bytesGuessed=0bytesAdded=0 entriesAdded=0 bytesRemoved=0 entriesRemoved=0
```

run 74, `tobias-1617787468-10-n4cpu4:2`

```
panic: TBG CRASH DETECTED GOID=168: r44: truncated cache to [2324, 2423] but got an entry for 2424..2435
add=false bytesGuessed=0bytesAdded=0 entriesAdded=0 bytesRemoved=0 entriesRemoved=0
```

run 110, `tobias-1617787468-09-n4cpu4:1`

```
panic: TBG CRASH DETECTED GOID=203: r40: truncated cache to [205, 304] but got an entry for 305..307
add=false bytesGuessed=0bytesAdded=0 entriesAdded=0 bytesRemoved=0 entriesRemoved=0
```

run 82, tobias-1617787468-50-n4cpu4

n1:

```
panic: TBG CRASH DETECTED GOID=194: r41: truncated cache to [2142, 2241] but got an entry for 2242..2256
add=false bytesGuessed=0bytesAdded=0 entriesAdded=0 bytesRemoved=0 entriesRemoved=0
```

n3
```
panic: TBG CRASH DETECTED GOID=149: r41: truncated cache to [2141, 2240] but got an entry for 2241..2249
add=false bytesGuessed=0bytesAdded=0 entriesAdded=0 bytesRemoved=0 entriesRemoved=0
```

run 89, `tobias-1617787468-25-n4cpu4:3`

```
panic: TBG CRASH DETECTED GOID=162: r40: truncated cache to [170, 269] but got an entry for 270..276
add=false bytesGuessed=0bytesAdded=0 entriesAdded=0 bytesRemoved=0 entriesRemoved=0
```

run 114, tobias-1617787468-46-n4cpu4:2

```
panic: TBG CRASH DETECTED GOID=158: r40: truncated cache to [143, 207] but got an entry for 208..263
add=false bytesGuessed=0bytesAdded=0 entriesAdded=0 bytesRemoved=0 entriesRemoved=0
```

run 126, tobias-1617787468-24-n4cpu4:2

```
panic: TBG CRASH DETECTED GOID=159: TBG term regression: idx 312 at term=8 appended with logterm=10
```

I'll leave this running but you get the idea. We catch the raft entry cache assertion quite a bit, and the term regression sometimes.

These repros all have `add=false`, which means that `(*Cache).Add` becomes really trivial - it basically just does

```go
	if truncate {
		truncIdx := ents[0].Index
		if add {
			// Some entries were already overwritten.
			truncIdx = ents[len(ents)-1].Index + 1
		}
		bytesRemoved, entriesRemoved = p.truncateFrom(truncIdx)
	}
```

I have the Cache here - should be doable to figure out why this didn't do what was advertised.

One last failure that occurred before I aborted the repro attempts:

run 144, `tobias-1617787468-18-n4cpu4:1`

```
panic: TBG CRASH DETECTED GOID=221: r41: truncated cache to [2429, 2551] but got an entry for 2552..2555
add=false bytesGuessed=0bytesAdded=0 entriesAdded=0 bytesRemoved=0 entriesRemoved=0
```

--- a/pkg/kv/kvserver/raftentry/BUILD.bazel
+++ b/pkg/kv/kvserver/raftentry/BUILD.bazel
@@ -22,7 +22,10 @@ go_library(
 go_test(
     name = "raftentry_test",
     size = "small",
-    srcs = ["cache_test.go"],
+    srcs = [
+        "cache_test.go",
+        "ring_buffer_test.go",
+    ],
     embed = [":raftentry"],
     deps = [
         "//pkg/roachpb",

--- a/pkg/kv/kvserver/raftentry/BUILD.bazel
+++ b/pkg/kv/kvserver/raftentry/BUILD.bazel
@@ -26,6 +26,7 @@ go_test(
     deps = [
         "//pkg/roachpb",
         "//pkg/util/leaktest",
+        "@com_github_stretchr_testify//require",
         "@io_etcd_go_etcd_raft_v3//raftpb",
     ],
 )

--- a/pkg/kv/kvserver/raftentry/BUILD.bazel
+++ b/pkg/kv/kvserver/raftentry/BUILD.bazel
@@ -11,6 +11,7 @@ go_library(
     visibility = ["//visibility:public"],
     deps = [
         "//pkg/roachpb",
+        "//pkg/util",
         "//pkg/util/metric",
         "//pkg/util/syncutil",
         "@com_github_cockroachdb_errors//:errors",

--- a/pkg/kv/kvserver/raftentry/cache.go
+++ b/pkg/kv/kvserver/raftentry/cache.go
@@ -180,19 +180,19 @@ func (c *Cache) Add(id roachpb.RangeID, ents []raftpb.Entry, truncate bool) {
 	p.mu.Lock()
 	defer p.mu.Unlock()
 	var bytesAdded, entriesAdded, bytesRemoved, entriesRemoved int32
-	if add {
-		bytesAdded, entriesAdded = p.add(ents)
-	}
+	// We truncate (if requested) before adding. This can lead to "wasted"
+	// work where we're zeroing out entries we would repopulate anyway, but
+	// past experience shows that it's best to keep this code as simple as
+	// possible (see #61990).
 	if truncate {
-		// Note that if !add, ents[0].Index may not even be in the cache
+		// Note that ents[0].Index may not even be in the cache
 		// at this point. `truncateFrom` will still remove any entries
 		// it may have at indexes >= truncIdx, as instructed.
 		truncIdx := ents[0].Index
-		if add {
-			// Some entries were already overwritten.
-			truncIdx = ents[len(ents)-1].Index + 1
-		}
 		bytesRemoved, entriesRemoved = p.truncateFrom(truncIdx)
+	}
+	if add {
+		bytesAdded, entriesAdded = p.add(ents)
 	}
 	c.recordUpdate(p, bytesAdded-bytesRemoved, bytesGuessed, entriesAdded-entriesRemoved)
 }
@@ -375,10 +375,10 @@ func analyzeEntries(ents []raftpb.Entry) (size int32) {
 	var prevTerm uint64
 	for i, e := range ents {
 		if i != 0 && e.Index != prevIndex+1 {
-			panic(errors.Errorf("invalid non-contiguous set of entries %d and %d", prevIndex, e.Index))
+			panic(errors.AssertionFailedf("invalid non-contiguous set of entries %d and %d", prevIndex, e.Index))
 		}
 		if i != 0 && e.Term < prevTerm {
-			err := errors.Errorf("term regression idx %d: %d -> %d", prevIndex, e.Term, prevTerm)
+			err := errors.AssertionFailedf("term regression idx %d: %d -> %d", prevIndex, prevTerm, e.Term)
 			panic(err)
 		}
 		prevIndex = e.Index

--- a/pkg/kv/kvserver/raftentry/cache.go
+++ b/pkg/kv/kvserver/raftentry/cache.go
@@ -184,6 +184,9 @@ func (c *Cache) Add(id roachpb.RangeID, ents []raftpb.Entry, truncate bool) {
 		bytesAdded, entriesAdded = p.add(ents)
 	}
 	if truncate {
+		// Note that if !add, ents[0].Index may not even be in the cache
+		// at this point. `truncateFrom` will still remove any entries
+		// it may have at indexes >= truncIdx, as instructed.
 		truncIdx := ents[0].Index
 		if add {
 			// Some entries were already overwritten.

--- a/pkg/kv/kvserver/raftentry/cache.go
+++ b/pkg/kv/kvserver/raftentry/cache.go
@@ -372,11 +372,17 @@ func (p *partition) setSize(orig, new cacheSize) bool {
 // entries in ents have contiguous indices.
 func analyzeEntries(ents []raftpb.Entry) (size int32) {
 	var prevIndex uint64
+	var prevTerm uint64
 	for i, e := range ents {
 		if i != 0 && e.Index != prevIndex+1 {
 			panic(errors.Errorf("invalid non-contiguous set of entries %d and %d", prevIndex, e.Index))
 		}
+		if i != 0 && e.Term < prevTerm {
+			err := errors.Errorf("term regression idx %d: %d -> %d", prevIndex, e.Term, prevTerm)
+			panic(err)
+		}
 		prevIndex = e.Index
+		prevTerm = e.Term
 		size += int32(e.Size())
 	}
 	return

--- a/pkg/kv/kvserver/raftentry/cache_test.go
+++ b/pkg/kv/kvserver/raftentry/cache_test.go
@@ -27,13 +27,28 @@ import (
 const noLimit = math.MaxUint64
 
 func newEntry(index, size uint64) raftpb.Entry {
+	r := rand.New(rand.NewSource(int64(index * size)))
 	data := make([]byte, size)
-	if _, err := rand.Read(data); err != nil {
+	if _, err := r.Read(data); err != nil {
 		panic(err)
 	}
-	return raftpb.Entry{
+	ent := raftpb.Entry{
 		Index: index,
 		Data:  data,
+	}
+	for {
+		entSize := uint64(ent.Size())
+		if entSize == size {
+			return ent
+		}
+		if entSize < size {
+			panic("size undershot")
+		}
+		delta := entSize - size
+		if uint64(len(ent.Data)) < delta {
+			panic("can't shorten ent.Data to target size")
+		}
+		ent.Data = ent.Data[delta:]
 	}
 }
 
@@ -46,7 +61,7 @@ func newEntries(lo, hi, size uint64) []raftpb.Entry {
 }
 
 func addEntries(c *Cache, rangeID roachpb.RangeID, lo, hi uint64) []raftpb.Entry {
-	ents := newEntries(lo, hi, 1)
+	ents := newEntries(lo, hi, 9)
 	c.Add(rangeID, ents, false)
 	return ents
 }
@@ -83,6 +98,26 @@ func verifyGet(
 			t.Fatalf("expected entry %v, but got %v", e, found)
 		}
 	}
+}
+
+func requireEqual(t *testing.T, c *Cache, rangeID roachpb.RangeID, idxs ...uint64) {
+	t.Helper()
+	p := c.getPartLocked(rangeID, false /* create */, false /* recordUse */)
+	if p == nil {
+		if len(idxs) > 0 {
+			t.Fatalf("expected idxs=%v but got empty cache", idxs)
+		}
+		return
+	}
+	b := &p.ringBuf
+	it := first(b)
+	var act []uint64
+	ok := it.valid(b)
+	for ok {
+		act = append(act, it.index(b))
+		it, ok = it.next(b)
+	}
+	require.Equal(t, idxs, act)
 }
 
 func TestEntryCache(t *testing.T) {
@@ -172,32 +207,47 @@ func TestIgnoredAdd(t *testing.T) {
 		require.True(t, 42*ignoredEnts[0].Size() > int(c.maxBytes)) // sanity check
 	}
 	verifyGet(t, c, rangeID, 1, 41, nil, 1, false)
+	requireEqual(t, c, rangeID)
 	verifyMetrics(t, c, 0, 0)
 	// Add some entries so we can show that a non-overlapping add is ignored.
 	ents := addEntries(c, rangeID, 4, 7)
 	verifyGet(t, c, rangeID, 4, 7, ents, 7, false)
+	requireEqual(t, c, rangeID, 4, 5, 6)
 	verifyMetrics(t, c, 3, 27+int64(partitionSize))
-	addEntries(c, rangeID, 1, 3)
+	addEntries(c, rangeID, 1, 3) // no-op because [1,2] does not overlap [4,5,6]
+	requireEqual(t, c, rangeID, 4, 5, 6)
 	verifyMetrics(t, c, 3, 27+int64(partitionSize))
 
 	// Cache has entries 4, 5, 6. Offer an oversize entry at index 7 (which is
 	// notably after 6) and request truncation. This should be a no-op.
 	c.Add(rangeID, []raftpb.Entry{newEntry(7, uint64(c.maxBytes+1))}, true /* truncate */)
+	requireEqual(t, c, rangeID, 4, 5, 6)
 	verifyGet(t, c, rangeID, 4, 7, ents, 7, false)
 
 	// Cache has entries 4, 5, 6. Offer an oversize entry at index 6 and request
 	// truncation. This should remove index 6 (as requested due to the truncation)
 	// without replacing it with the input entry.
 	c.Add(rangeID, []raftpb.Entry{newEntry(6, uint64(c.maxBytes+1))}, true /* truncate */)
+	requireEqual(t, c, rangeID, 4, 5)
 	verifyGet(t, c, rangeID, 4, 7, ents[:len(ents)-1], 6, false)
 
 	// Cache has entries 4, 5. Offer an oversize entry at index 3 (which is
 	// notably before 4) and request truncation. This should clear all entries
-	//>= 3, i.e. everything.
+	// >= 3, i.e. everything.
 	c.Add(rangeID, []raftpb.Entry{newEntry(3, uint64(c.maxBytes+1))}, true /* truncate */)
 	// And it did.
+	requireEqual(t, c, rangeID)
 	verifyGet(t, c, rangeID, 0, 0, nil, 0, false)
 	verifyMetrics(t, c, 0, int64(partitionSize))
+
+	addEntries(c, rangeID, 10, 13)
+	// Now, cache = [10, 11, 12].
+	requireEqual(t, c, rangeID, 10, 11, 12)
+	verifyMetrics(t, c, 3, 3*9+int64(partitionSize))
+
+	c.Add(rangeID, newEntries(3, 4, 9), true /* truncate */)
+	requireEqual(t, c, rangeID, 3)
+	verifyMetrics(t, c, 1, 1*9+int64(partitionSize))
 }
 
 func TestRingBuffer_truncateFrom(t *testing.T) {
@@ -207,7 +257,7 @@ func TestRingBuffer_truncateFrom(t *testing.T) {
 	const maxBytes = 100
 	c := NewCache(maxBytes)
 	// Add one entry.
-	c.Add(rangeID, newEntries(100, 101, 0), false /* truncate */)
+	c.Add(rangeID, newEntries(100, 101, 9), false /* truncate */)
 	ents, _, _, _ := c.Scan(nil, rangeID, 100, 101, noLimit)
 	// Entry is actually there.
 	require.Len(t, ents, 1)
@@ -229,7 +279,7 @@ func TestRingBuffer_clearTo(t *testing.T) {
 	const maxBytes = 100
 	c := NewCache(maxBytes)
 	// Add one entry.
-	c.Add(rangeID, newEntries(100, 101, 0), false /* truncate */)
+	c.Add(rangeID, newEntries(100, 101, 9), false /* truncate */)
 	ents, _, _, _ := c.Scan(nil, rangeID, 100, 101, noLimit)
 	// Entry is actually there.
 	require.Len(t, ents, 1)
@@ -256,7 +306,7 @@ func TestAddAndTruncate(t *testing.T) {
 	verifyMetrics(t, c, 6, 54+int64(partitionSize))
 	// Show that even if the addition is ignored due to size, entries
 	// with an equal or larger index are truncated.
-	largeEnts := newEntries(5, 6, 300)
+	largeEnts := newEntries(5, 6, 900)
 	c.Add(rangeID, largeEnts, true /* truncate */)
 	verifyGet(t, c, rangeID, 1, 10, ents[:4], 5, false)
 	verifyMetrics(t, c, 4, 36+int64(partitionSize))
@@ -269,8 +319,8 @@ func TestDrop(t *testing.T) {
 		r2 roachpb.RangeID = 2
 
 		sizeOf9Entries = 81
-		partitionSize  = int64(sizeOf9Entries + partitionSize)
 	)
+	partitionSize := int64(sizeOf9Entries + partitionSize)
 	c := NewCache(1 << 10)
 	ents1 := addEntries(c, r1, 1, 10)
 	verifyGet(t, c, r1, 1, 10, ents1, 10, false)
@@ -314,7 +364,7 @@ func TestEntryCacheClearTo(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	rangeID := roachpb.RangeID(1)
 	c := NewCache(100)
-	c.Add(rangeID, []raftpb.Entry{newEntry(20, 1), newEntry(21, 1)}, true)
+	c.Add(rangeID, []raftpb.Entry{newEntry(20, 9), newEntry(21, 9)}, true)
 	c.Clear(rangeID, 21)
 	c.Clear(rangeID, 18)
 	if ents, _, _, _ := c.Scan(nil, rangeID, 2, 21, noLimit); len(ents) != 0 {
@@ -389,12 +439,12 @@ func TestConcurrentEvictions(t *testing.T) {
 		if offset >= 0 && offset < len(ents) {
 			lo := ents[offset].Index
 			hi := lo + uint64(length)
-			toAdd = newEntries(lo, hi, 1)
+			toAdd = newEntries(lo, hi, 9)
 			ents = append(ents[:offset], toAdd...)
 		} else {
 			lo := uint64(offset + 2)
 			hi := lo + uint64(length)
-			toAdd = newEntries(lo, hi, 1)
+			toAdd = newEntries(lo, hi, 9)
 			ents = toAdd
 		}
 		rangeData[r] = ents
@@ -483,7 +533,7 @@ func TestEntryCacheEviction(t *testing.T) {
 	}
 	// Add another entry to the same range. This will exceed the size limit and
 	// lead to eviction.
-	c.Add(rangeID, []raftpb.Entry{newEntry(3, 40)}, true)
+	c.Add(rangeID, []raftpb.Entry{newEntry(3, 80)}, true)
 	ents, _, hi, _ = c.Scan(nil, rangeID, 1, 4, noLimit)
 	if len(ents) != 0 || hi != 1 {
 		t.Errorf("expected no entries; got %+v, %d", ents, hi)
@@ -498,9 +548,9 @@ func TestEntryCacheEviction(t *testing.T) {
 	if len(ents) != 1 || hi != 4 {
 		t.Errorf("expected the new entry; got %+v, %d", ents, hi)
 	}
-	c.Add(rangeID, []raftpb.Entry{newEntry(3, 1)}, true)
+	c.Add(rangeID, []raftpb.Entry{newEntry(3, 9)}, true)
 	verifyMetrics(t, c, 1, c.Metrics().Bytes.Value())
-	c.Add(rangeID2, []raftpb.Entry{newEntry(20, 1), newEntry(21, 1)}, true)
+	c.Add(rangeID2, []raftpb.Entry{newEntry(20, 9), newEntry(21, 9)}, true)
 	ents, _, hi, _ = c.Scan(nil, rangeID2, 20, 22, noLimit)
 	if len(ents) != 2 || hi != 22 {
 		t.Errorf("expected both entries; got %+v, %d", ents, hi)
@@ -612,7 +662,7 @@ func BenchmarkEntryCache(b *testing.B) {
 	rangeID := roachpb.RangeID(1)
 	ents := make([]raftpb.Entry, 1000)
 	for i := range ents {
-		ents[i] = newEntry(uint64(i+1), 8)
+		ents[i] = newEntry(uint64(i+1), 9)
 	}
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
@@ -634,7 +684,7 @@ func BenchmarkEntryCacheClearTo(b *testing.B) {
 	rangeID := roachpb.RangeID(1)
 	ents := make([]raftpb.Entry, 1000)
 	for i := range ents {
-		ents[i] = newEntry(uint64(i+1), 8)
+		ents[i] = newEntry(uint64(i+1), 9)
 	}
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {

--- a/pkg/kv/kvserver/raftentry/cache_test.go
+++ b/pkg/kv/kvserver/raftentry/cache_test.go
@@ -20,6 +20,7 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/stretchr/testify/require"
 	"go.etcd.io/etcd/raft/v3/raftpb"
 )
 
@@ -166,7 +167,10 @@ func TestIgnoredAdd(t *testing.T) {
 	rangeID := roachpb.RangeID(1)
 	c := NewCache(100 + uint64(partitionSize))
 	// Show that adding entries which are larger than maxBytes is ignored.
-	_ = addEntries(c, rangeID, 1, 41)
+	{
+		ignoredEnts := addEntries(c, rangeID, 1, 41)
+		require.True(t, 42*ignoredEnts[0].Size() > int(c.maxBytes)) // sanity check
+	}
 	verifyGet(t, c, rangeID, 1, 41, nil, 1, false)
 	verifyMetrics(t, c, 0, 0)
 	// Add some entries so we can show that a non-overlapping add is ignored.
@@ -175,6 +179,67 @@ func TestIgnoredAdd(t *testing.T) {
 	verifyMetrics(t, c, 3, 27+int64(partitionSize))
 	addEntries(c, rangeID, 1, 3)
 	verifyMetrics(t, c, 3, 27+int64(partitionSize))
+
+	// Cache has entries 4, 5, 6. Offer an oversize entry at index 7 (which is
+	// notably after 6) and request truncation. This should be a no-op.
+	c.Add(rangeID, []raftpb.Entry{newEntry(7, uint64(c.maxBytes+1))}, true /* truncate */)
+	verifyGet(t, c, rangeID, 4, 7, ents, 7, false)
+
+	// Cache has entries 4, 5, 6. Offer an oversize entry at index 6 and request
+	// truncation. This should remove index 6 (as requested due to the truncation)
+	// without replacing it with the input entry.
+	c.Add(rangeID, []raftpb.Entry{newEntry(6, uint64(c.maxBytes+1))}, true /* truncate */)
+	verifyGet(t, c, rangeID, 4, 7, ents[:len(ents)-1], 6, false)
+
+	// Cache has entries 4, 5. Offer an oversize entry at index 3 (which is
+	// notably before 4) and request truncation. This should clear all entries
+	//>= 3, i.e. everything.
+	c.Add(rangeID, []raftpb.Entry{newEntry(3, uint64(c.maxBytes+1))}, true /* truncate */)
+	// And it did.
+	verifyGet(t, c, rangeID, 0, 0, nil, 0, false)
+	verifyMetrics(t, c, 0, int64(partitionSize))
+}
+
+func TestRingBuffer_truncateFrom(t *testing.T) {
+	// NB: this is also exercised through the truncate=true test cases in
+	// TestIgnoredAdd, but this tests it more directly.
+	rangeID := roachpb.RangeID(1)
+	const maxBytes = 100
+	c := NewCache(maxBytes)
+	// Add one entry.
+	c.Add(rangeID, newEntries(100, 101, 0), false /* truncate */)
+	ents, _, _, _ := c.Scan(nil, rangeID, 100, 101, noLimit)
+	// Entry is actually there.
+	require.Len(t, ents, 1)
+	p := c.getPartLocked(rangeID, false /* create */, false /* recordUse */)
+	require.NotNil(t, p)
+	// Truncate range [99, infinity], which should work even though
+	// 99 isn't itself in `p`.
+	_, numRemovedEntries := p.truncateFrom(99)
+	require.EqualValues(t, 1, numRemovedEntries)
+	require.Zero(t, p.len)
+	ents, _, _, _ = c.Scan(nil, rangeID, 100, 101, noLimit)
+	require.Empty(t, ents)
+}
+
+func TestRingBuffer_clearTo(t *testing.T) {
+	// Ensure that clearTo isn't sensitive to whether its argument actually
+	// refers to a cached index.
+	rangeID := roachpb.RangeID(1)
+	const maxBytes = 100
+	c := NewCache(maxBytes)
+	// Add one entry.
+	c.Add(rangeID, newEntries(100, 101, 0), false /* truncate */)
+	ents, _, _, _ := c.Scan(nil, rangeID, 100, 101, noLimit)
+	// Entry is actually there.
+	require.Len(t, ents, 1)
+	p := c.getPartLocked(rangeID, false /* create */, false /* recordUse */)
+	require.NotNil(t, p)
+	_, numRemovedEntries := p.clearTo(101)
+	require.EqualValues(t, 1, numRemovedEntries)
+	require.Zero(t, p.len)
+	ents, _, _, _ = c.Scan(nil, rangeID, 100, 101, noLimit)
+	require.Empty(t, ents)
 }
 
 func TestAddAndTruncate(t *testing.T) {

--- a/pkg/kv/kvserver/raftentry/ring_buffer.go
+++ b/pkg/kv/kvserver/raftentry/ring_buffer.go
@@ -62,9 +62,17 @@ func (b *ringBuf) add(ents []raftpb.Entry) (addedBytes, addedEntries int32) {
 
 // truncateFrom clears all entries from the ringBuf with index equal to or
 // greater than lo. The method returns the aggregate size and count of entries
-// removed.
+// removed. Note that lo itself may or may not be in the cache.
 func (b *ringBuf) truncateFrom(lo uint64) (removedBytes, removedEntries int32) {
 	it, ok := iterateFrom(b, lo)
+	if !ok {
+		if first(b).index(b) > lo {
+			// If `lo` precedes the indexes in the buffer
+			// (say the buf is idx=[100, 101, 102] and `lo` is 99),
+			// we need to truncate everything.
+			it, ok = iterateFrom(b, first(b).index(b))
+		}
+	}
 	for ok {
 		removedBytes += int32(it.entry(b).Size())
 		removedEntries++

--- a/pkg/kv/kvserver/raftentry/ring_buffer.go
+++ b/pkg/kv/kvserver/raftentry/ring_buffer.go
@@ -34,10 +34,10 @@ const (
 // given that ents may overlap with existing entries or may be rejected from
 // the buffer. ents must not be empty.
 func (b *ringBuf) add(ents []raftpb.Entry) (addedBytes, addedEntries int32) {
-	if afterCache := b.len > 0 && ents[0].Index > last(b).index(b)+1; afterCache {
+	if it := last(b); it.valid(b) && ents[0].Index > it.index(b)+1 {
 		// If ents is non-contiguous and later than the currently cached range then
 		// remove the current entries and add ents in their place.
-		removedBytes, removedEntries := b.clearTo(last(b).index(b) + 1)
+		removedBytes, removedEntries := b.clearTo(it.index(b) + 1)
 		addedBytes, addedEntries = -1*removedBytes, -1*removedEntries
 	}
 	before, after, ok := computeExtension(b, ents[0].Index, ents[len(ents)-1].Index)
@@ -66,15 +66,18 @@ func (b *ringBuf) add(ents []raftpb.Entry) (addedBytes, addedEntries int32) {
 // greater than lo. The method returns the aggregate size and count of entries
 // removed. Note that lo itself may or may not be in the cache.
 func (b *ringBuf) truncateFrom(lo uint64) (removedBytes, removedEntries int32) {
-	it, ok := iterateFrom(b, lo)
-	if !ok {
-		if first(b).index(b) > lo {
-			// If `lo` precedes the indexes in the buffer
-			// (say the buf is idx=[100, 101, 102] and `lo` is 99),
-			// we need to truncate everything.
-			it, ok = iterateFrom(b, first(b).index(b))
-		}
+	if b.len == 0 {
+		return
 	}
+	if idx := first(b).index(b); idx > lo {
+		// If `lo` precedes the indexes in the buffer
+		// (say the buf is idx=[100, 101, 102] and `lo` is 99),
+		// `iterateFrom` will return an invalid iter. But we
+		// need to truncate everything and so advance to the
+		// first index before constructing the iterator.
+		lo = idx
+	}
+	it, ok := iterateFrom(b, lo)
 	for ok {
 		removedBytes += int32(it.entry(b).Size())
 		removedEntries++
@@ -88,15 +91,14 @@ func (b *ringBuf) truncateFrom(lo uint64) (removedBytes, removedEntries int32) {
 	if util.RaceEnabled {
 		if b.len > 0 {
 			if lastIdx := last(b).index(b); lastIdx >= lo {
-				err := errors.Errorf(
+				panic(errors.AssertionFailedf(
 					"buffer truncated to [..., %d], but current last index is %d",
 					lo, lastIdx,
-				)
-				panic(err)
+				))
 			}
 		}
 	}
-	return
+	return removedBytes, removedEntries
 }
 
 // clearTo clears all entries from the ringBuf with index less than hi. The
@@ -105,7 +107,8 @@ func (b *ringBuf) clearTo(hi uint64) (removedBytes, removedEntries int32) {
 	if b.len == 0 || hi < first(b).index(b) {
 		return
 	}
-	it, ok := first(b), true
+	it := first(b)
+	ok := it.valid(b) // true
 	firstIndex := it.index(b)
 	for ok && it.index(b) < hi {
 		removedBytes += int32(it.entry(b).Size())
@@ -234,12 +237,18 @@ func iterateFrom(b *ringBuf, index uint64) (_ iterator, ok bool) {
 // first returns an iterator pointing to the first entry of the ringBuf.
 // If b is empty, the returned iterator is not valid.
 func first(b *ringBuf) iterator {
+	if b.len == 0 {
+		return iterator(-1)
+	}
 	return iterator(b.head)
 }
 
 // last returns an iterator pointing to the last element in b.
-// It is unsafe to call last if b has an empty buffer.
+// If b is empty, the returned iterator is not valid.
 func last(b *ringBuf) iterator {
+	if b.len == 0 {
+		return iterator(-1)
+	}
 	return iterator((b.head + b.len - 1) % len(b.buf))
 }
 

--- a/pkg/kv/kvserver/raftentry/ring_buffer.go
+++ b/pkg/kv/kvserver/raftentry/ring_buffer.go
@@ -13,6 +13,8 @@ package raftentry
 import (
 	"math/bits"
 
+	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/cockroachdb/errors"
 	"go.etcd.io/etcd/raft/v3/raftpb"
 )
 
@@ -82,6 +84,17 @@ func (b *ringBuf) truncateFrom(lo uint64) (removedBytes, removedEntries int32) {
 	b.len -= int(removedEntries)
 	if b.len < (len(b.buf) / shrinkThreshold) {
 		realloc(b, 0, b.len)
+	}
+	if util.RaceEnabled {
+		if b.len > 0 {
+			if lastIdx := last(b).index(b); lastIdx >= lo {
+				err := errors.Errorf(
+					"buffer truncated to [..., %d], but current last index is %d",
+					lo, lastIdx,
+				)
+				panic(err)
+			}
+		}
 	}
 	return
 }

--- a/pkg/kv/kvserver/raftentry/ring_buffer_test.go
+++ b/pkg/kv/kvserver/raftentry/ring_buffer_test.go
@@ -1,0 +1,215 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package raftentry
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestRingBuffer_Add(t *testing.T) {
+	b := &ringBuf{}
+	const size = 12
+	b.add(newEntries(5, 8, size))
+	eq(t, b, 5, 6, 7)
+	b.add(newEntries(3, 5, size))
+	eq(t, b, 3, 4, 5, 6, 7)
+
+	{
+		// Overwrite 5 and 6 with entries twice as large.
+		ab, ae := b.add(newEntries(5, 7, 2*size))
+		eq(t, b, 3, 4, 5, 6, 7)
+		require.EqualValues(t, 2*size, ab) // `size` for idx=5 and 6
+		require.EqualValues(t, 0, ae)      // we overwrote, did not add
+	}
+
+	{
+		rb, re := b.truncateFrom(6) // reduce typing work below
+		eq(t, b, 3, 4, 5)
+		require.EqualValues(t, 3*size, rb) // idx=6,7 have double size
+		require.EqualValues(t, 2, re)
+	}
+
+	{
+		// Add one entry at the end.
+		ab, ae := b.add(newEntries(6, 7, size))
+		require.EqualValues(t, size, ab)
+		require.EqualValues(t, 1, ae)
+		eq(t, b, 3, 4, 5, 6)
+	}
+
+	{
+		// Addition at beginning which matches up is accepted.
+		ab, ae := b.add(newEntries(1, 3, size))
+		eq(t, b, 1, 2, 3, 4, 5, 6)
+		require.EqualValues(t, 2*size, ab)
+		require.EqualValues(t, 2, ae)
+	}
+
+	{
+		rb, re := b.clearTo(3)
+		eq(t, b, 3, 4, 5, 6)
+		require.EqualValues(t, rb, 2*size)
+		require.EqualValues(t, re, 2)
+	}
+
+	{
+		// Addition at beginning which does not line up is rejected.
+		ab, ae := b.add(newEntries(1, 2, size))
+		eq(t, b, 3, 4, 5, 6)
+		require.Zero(t, ab)
+		require.Zero(t, ae)
+	}
+
+	{
+		// Addition at the end but with a gap clears the existing entries
+		// before.
+		b.add(newEntries(10, 11, size))
+		eq(t, b, 10)
+	}
+}
+
+func TestRingBuffer_Scan(t *testing.T) {
+	for _, tc := range []struct {
+		desc       string
+		lo, hi, mb uint64
+
+		idxs             []uint64 // full range is 10,11,12,13,14, each of size 10
+		scanBytes        uint64
+		nextIdx          uint64
+		exceededMaxBytes bool
+	}{
+		{
+			desc: "before cached entries",
+			lo:   5, hi: 10, mb: 100,
+			nextIdx: 5,
+		},
+		{
+			desc: "before up to excluding first cached entry",
+			lo:   5, hi: 11, mb: 100,
+			nextIdx: 5,
+		},
+		{
+			desc: "starts at first cached entry, remains in cache",
+			lo:   10, hi: 12, mb: 100,
+			idxs: []uint64{10, 11}, scanBytes: 20, nextIdx: 12,
+		},
+		{
+			desc: "starts at first cached entry, remains in cache, limit almost hit",
+			lo:   10, hi: 12, mb: 20,
+			idxs: []uint64{10, 11}, scanBytes: 20, nextIdx: 12, exceededMaxBytes: false,
+		},
+		{
+			desc: "starts at first cached entry, remains in cache, limit hit",
+			lo:   10, hi: 12, mb: 19,
+			idxs: []uint64{10}, scanBytes: 10, nextIdx: 11, exceededMaxBytes: true,
+		},
+		{
+			desc: "starts at first cached entry, stops at cache end",
+			lo:   10, hi: 15, mb: 100,
+			idxs: []uint64{10, 11, 12, 13, 14}, scanBytes: 50, nextIdx: 15,
+		},
+		{
+			desc: "starts at first cached entry, runs past the cache",
+			lo:   10, hi: 16, mb: 100,
+			idxs: []uint64{10, 11, 12, 13, 14}, scanBytes: 50, nextIdx: 15,
+		},
+		{
+			desc: "starts in middle of cache, runs past the cache",
+			lo:   12, hi: 16, mb: 100,
+			idxs: []uint64{12, 13, 14}, scanBytes: 30, nextIdx: 15,
+		},
+		{
+			desc: "starts in middle of cache, runs past the cache but limit hits",
+			lo:   12, hi: 16, mb: 29,
+			idxs: []uint64{12, 13}, scanBytes: 20, nextIdx: 14, exceededMaxBytes: true,
+		},
+		{
+			desc: "starts past cache",
+			lo:   15, hi: 16, mb: 100,
+			nextIdx: 15,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			b := &ringBuf{}
+			ab, _ := b.add(newEntries(10, 15, 10 /* size */))
+			require.EqualValues(t, 5*10, ab)
+			ents, sb, next, excmb := b.scan(nil, tc.lo, tc.hi, tc.mb)
+			var sl []uint64
+			for _, ent := range ents {
+				sl = append(sl, ent.Index)
+			}
+			require.Equal(t, tc.idxs, sl)
+			require.Equal(t, tc.scanBytes, sb)
+			require.Equal(t, tc.nextIdx, next)
+			require.Equal(t, tc.exceededMaxBytes, excmb)
+		})
+	}
+}
+
+func TestRingBuffer_TruncateFrom(t *testing.T) {
+	b := &ringBuf{}
+	b.truncateFrom(20)
+	eq(t, b)
+	b.add(newEntries(10, 20, 9))
+	eq(t, b, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19)
+	b.truncateFrom(20)
+	eq(t, b, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19)
+	b.truncateFrom(19)
+	eq(t, b, 10, 11, 12, 13, 14, 15, 16, 17, 18)
+	b.truncateFrom(13)
+	eq(t, b, 10, 11, 12)
+	b.truncateFrom(1)
+	eq(t, b)
+	it := first(b)
+	require.False(t, it.valid(b)) // regression test
+}
+
+func TestRingBuffer_ClearTo(t *testing.T) {
+	b := &ringBuf{}
+	b.clearTo(100)
+	eq(t, b)
+	b.add(newEntries(10, 20, 9))
+	eq(t, b, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19)
+	b.clearTo(10)
+	eq(t, b, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19)
+	b.clearTo(11)
+	eq(t, b, 11, 12, 13, 14, 15, 16, 17, 18, 19)
+	b.clearTo(17)
+	eq(t, b, 17, 18, 19)
+	b.clearTo(100)
+	eq(t, b)
+}
+
+func eq(t *testing.T, b *ringBuf, idxs ...uint64) {
+	t.Helper()
+	var sl []uint64
+	it := first(b)
+	for it.valid(b) {
+		idx := it.index(b)
+		sl = append(sl, idx)
+		it, _ = it.next(b)
+		ent, ok := b.get(idx)
+		require.True(t, ok)
+		require.Equal(t, idx, ent.Index)
+	}
+	require.Equal(t, idxs, sl)
+	if len(sl) == 0 {
+		return
+	}
+	// NB: this sufficiently tests `Get`, so it doesn't have its own
+	// unit tests.
+	_, ok := b.get(sl[0] - 1)
+	require.False(t, ok)
+	_, ok = b.get(sl[len(sl)-1] + 1)
+	require.False(t, ok)
+}

--- a/pkg/kv/kvserver/replica_raft.go
+++ b/pkg/kv/kvserver/replica_raft.go
@@ -1131,8 +1131,33 @@ func (r *Replica) sendRaftMessages(ctx context.Context, messages []raftpb.Messag
 				// contain fat entries. Since these are the only two sources that
 				// raft.sendAppend gathers entries from to populate MsgApps, we
 				// should never see thin entries here.
+				//
+				// Also assert that the log term only ever increases (most of the
+				// time it stays constant, as term changes are rare), and that
+				// the index increases by exactly one with each entry.
+				//
+				// This assertion came out of #61990.
+				prevTerm := message.LogTerm // term of entry preceding the append
+				prevIndex := message.Index  // index of entry preceding the append
 				for j := range message.Entries {
-					assertSideloadedRaftCommandInlined(ctx, &message.Entries[j])
+					ent := &message.Entries[j]
+					assertSideloadedRaftCommandInlined(ctx, ent)
+
+					if prevIndex+1 != ent.Index {
+						log.Fatalf(ctx,
+							"index gap in outgoing MsgApp: idx %d followed by %d",
+							prevIndex, ent.Index,
+						)
+					}
+					prevIndex = ent.Index
+					if prevTerm > ent.Term {
+						log.Fatalf(ctx,
+							"term regression in outgoing MsgApp: idx %d at term=%d "+
+								"appended with logterm=%d",
+							ent.Index, ent.Term, message.LogTerm,
+						)
+					}
+					prevTerm = ent.Term
 				}
 			}
 

